### PR TITLE
feat(typeahead): add demo and property for adding optionsListTemplate 

### LINKED ISSUE #4036: Typeaheads' optionsListTemplate does not provide full component context
### Bug description or feature request:

**Bug** When using Typeahead's optionsListTemplate option, the full component context is not passed into the specified template.  Instead only matches, itemTemplate and query are passed in.

This means you can't replicate basic functionality that the default template provides (such as selecting an item, highlighting the active item, etc) without duplicating the existing logic into your own component.

I believe this should be able to be easily resolved by modifying typeahead-container.component.html and changing:
` [ngTemplateOutletContext]="{matches:matches, itemTemplate:itemTemplate, query:query}"`

to
`[ngTemplateOutletContext]="this"`

### Versions of ngx-bootstrap:
ngx-bootstrap:
2.0.3

### PR DESCRIPTION
**Pull request by ghost**
Add optionsListTemplate input to typeahead
Add a new demo for optionsListTemplate
Create models TypeaheadTemplateMethods, TypeaheadOptionItemContext, TypeaheadOptionListContext

Closes #4036, #3081

# PR Checklist
Before creating new PR, please take a look at checklist below to make sure that you've done everything that needs to be done before we can merge it.

 - [X] read and followed the [CONTRIBUTING.md](https://github.com/valor-software/ngx-bootstrap/blob/development/CONTRIBUTING.md) guide.
 - [X] built and tested the changes locally.
 - [ ] added/updated tests.
 - [x] added/updated API documentation.
 - [X] added/updated demos.

### CODE REVIEW

File: src/typeahead/typeahead-container.component.html [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <!-- inject options list template -->
 <ng-template [ngTemplateOutlet]="optionsListTemplate || (isBs4 ? bs4Template : bs3Template)"
-             [ngTemplateOutletContext]="{matches:matches, itemTemplate:itemTemplate, query:query}"></ng-template>
+             [ngTemplateOutletContext]="{matches:matches, itemTemplate:itemTemplate, query:query, $implicit: this, isBs4: this.isBs4}"></ng-template>
```
Comment (Domainv) [inline src/typeahead/typeahead-container.component.html:10]: I’m not sure about this solution. You need to throw more specific methods into the context.

**Review by Domainv — CHANGES_REQUESTED** [review]

File: src/typeahead/typeahead-container.component.html (hunk 1 above)
Comment (valorkin) [reply src/typeahead/typeahead-container.component.html:10]: replace `this` with an object or set of properties

**Review by valorkin — DISMISSED** [review]
one demo - one use case

**Review by Domainv — DISMISSED** [review]

File: src/typeahead/typeahead-container.component.ts [hunk 2]
```diff
@@ -130,7 +142,7 @@ export class TypeaheadContainerComponent {
     }
   }
 
-  selectActive(value: TypeaheadMatch): void {
+  selectActive(value: TypeaheadMatch, event?: any): void {
```
Comment (Domainv) [inline src/typeahead/typeahead-container.component.ts]: why do we need event there?

Comment (ghost) [reply src/typeahead/typeahead-container.component.ts]: removed it

**Review by Domainv — DISMISSED** [review]

**Comment by ludmilanesvitiy** [comment]
Tested new Demo typeahead, looks good.

File: src/typeahead/typeahead-container.component.ts [hunk 3]
```diff
@@ -45,6 +45,18 @@ export class TypeaheadContainerComponent {
     return !isBs3();
   }
 
+  get typeaheadTemplateMethods(): {} {
```
Comment (valorkin) [inline src/typeahead/typeahead-container.component.ts]: @svetoldo4444ka you forgot to add type

File: src/typeahead/typeahead-container.component.ts [hunk 4]
```diff
@@ -45,6 +45,18 @@ export class TypeaheadContainerComponent {
...
+    /* tslint:disable:no-this-assignment */
+    const _that = this;
+
+    return {
+      isBs4: _that.isBs4,
+      selectMatch: this.selectMatch.bind(_that),
```
Comment (valorkin) [inline src/typeahead/typeahead-container.component.ts:74]: I can smell react approach :)

**Review by daniloff200 — APPROVED** [review]

**Comment by dmitry-zhemchugov** [comment]
Tested. Ready to merge

**Pull request merged**